lib/system/arc.nim

```diff
@@ -116,15 +116,30 @@ proc nimNewObjUninit(size, alignment: int): pointer {.compilerRtl.} =
     cprintf("[Allocated] %p result: %p\n", result -! sizeof(RefHeader), result)
 
 proc nimDecWeakRef(p: pointer) {.compilerRtl, inl.} =
-  dec head(p).rc, rcIncrement
+  # We want to use atomic operations when threading is on
+  # We will only use non-atomic operations if user 'owns' the ref - which implies
+  # that no other thread is handling the ref anymore.
+  template decOpr(x,y: untyped): untyped =
+    when hasThreadSupport:
+      atomicDec(x, y, ATOMIC_ACQ_REL)
+    else:
+      dec(x, y)
```

Comment:
Move it out and reuse it below.  Add a comment or more descriptive name.  Why not use `typed`?